### PR TITLE
Optimize {Encode,Decode}Bytes by using bytes.IndexByte.

### PR DESCRIPTION
**Pull request by petermattis**
benchmark                old ns/op     new ns/op     delta
BenchmarkEncodeBytes     83.8          44.7          -46.66%
BenchmarkDecodeBytes     84.6          43.0          -49.17%

### CODE REVIEW

File: util/encoding/encoding.go [hunk 1]
```diff
@@ -370,25 +373,27 @@ func EncodeBytes(b []byte, data []byte) []byte {
...
+		}
+		if i+1 > len(b) {
+			panic("malformed escape")
+		}
+		v := b[i+1]
+		if v == escapedTerm {
+			if r == nil {
+				r = b[:i]
+			} else {
+				r = append(r, b[:i]...)
```
Comment (tschottdorf) [inline util/encoding/encoding.go]: this looks like it changes the input buffer, which I didn't expect based on the description of the function.

Comment (petermattis) [inline util/encoding/encoding.go]: Where are you seeing the input buffer being modified? It shouldn't be. The above line appends `b[:i]` to `r`. Sometimes we set `r` to `b[:i]`, but only when we can return a portion of the input buffer unmodified as the result.

File: util/encoding/encoding.go [hunk 2]
```diff
@@ -370,25 +373,27 @@ func EncodeBytes(b []byte, data []byte) []byte {
...
 			}
-			copyStart = i + 1
+			return b[i+2:], r
+		}
+		r = append(r, b[:i]...)
+		if v == escapedNul {
+			r = append(r, 0)
 		}
```
Comment (bdarnell) [inline util/encoding/encoding.go]: There should be a panic here when the escape character is followed by something other than escapedTerm or escapedNul (or we should add an error return and stop panicking)

**Comment by bdarnell** [comment]
LGTM

File: util/encoding/encoding.go (hunk 1 above)
Comment (tschottdorf) [inline util/encoding/encoding.go]: sorry, didn't see the return below. You're right of course.

**Comment by tschottdorf** [comment]
LGTM

**Pull request merged**